core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,320 @@
+            IndexedPriorityQueue<Integer> maxWriterPartitions,
+            Set<Integer> rebalancedPartitionIds)
+    {
+        while (true) {
+            // Find the partition with maximum row count
+            EntryWithPriority<Integer> partitionToScale = maxWriterPartitions.pollEntryWithPriority();
+
+            if (partitionToScale == null) {
+                return Optional.empty();
+            }
+
+            int partitionId = partitionToScale.value();
+            long rowCount = partitionToScale.priority();
+
+            PartitionWriterInfo partitionWriterInfo = partitionWriterInfos[partitionId];
+
+            // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+            // this partition and move on to the other partition inside max writer.
+            if (rebalancedPartitionIds.contains(partitionId) || partitionWriterInfo.containsWriter(minWriterId)) {
+                continue;
+            }
+
+            long estimatedPartitionWrittenBytes = (maxWriterWrittenBytesSinceLastRebalance * rowCount) / maxWriterRowCountSinceLastRebalance;
+
+            // Scale the partition when estimated physicalWrittenBytes is greater than writerMinSize.
+            if (partitionWriterInfo.getWriterCount() <= numberOfWriters && estimatedPartitionWrittenBytes >= writerMinSize) {
+                partitionWriterInfo.addWriter(minWriterId);
+                log.warn(format("Scaled partition (%s) to writer %s with writer count %s", partitionId, minWriterId, partitionWriterInfo.getWriterCount()));
+                return Optional.of(new IdWithWrittenBytes(partitionId, estimatedPartitionWrittenBytes));
+            }
+
+            return Optional.empty();
+        }
+    }
+
+    private void resetStateForNextRebalance(long[] writerPhysicalWrittenBytes)
+    {
+        for (int i = 0; i < numberOfWriters; i++) {
+            writerPartitionIdWithRowCount.set(i, new IndexedPriorityQueue<>());
+            lastRefreshWriterPhysicalWrittenBytes[i] = writerPhysicalWrittenBytes[i];
+        }
+    }
+
+    public synchronized void addPartitionRowCount(int writerId, int partitionId, long rowCount)
+    {
+        IndexedPriorityQueue<Integer> partitions = writerPartitionIdWithRowCount.get(writerId);
+        long newRowCount = partitions.getPriority(partitionId)
+                .map(currentRowCount -> currentRowCount + rowCount)
+                .orElse(rowCount);
+        partitions.addOrUpdate(partitionId, newRowCount);
+    }
+
+    public synchronized int getNextWriterId(int partitionId)
+    {
+        return partitionWriterInfos[partitionId].getNextWriterId();
+    }
+
+    @VisibleForTesting
+    synchronized List<Integer> getWriterIds(int partitionId)
+    {
+        return partitionWriterInfos[partitionId].getWriterIds();
+    }
+
+    private static class PartitionWriterInfo
+    {
+        private final List<Integer> writerIds = new ArrayList<>();
```

Comment:
nit: Maybe this should be a set?

Reply:
this can't be a set since we need to do round robin using list

Reply:
Ok. There is for example `LinkedHashSet`